### ISSUE TITLE
What is metadata?

### ISSUE THREAD

**Issue by jtremback** [paritytech/jsonrpc#179]
I don't see any documentation here or in the jsonrpc spec on "metadata", and what its role is.

**Comment by tomusdrw** [comment]
`Metadata` is a custom extension of JSONRPC spec. It allows a method handler to access some additional data from the transport layer, which is not part of the payload itself.

For instance you can check a HTTP header and alter the behaviour (obviously it's against transport-agnostic JSONRPC spec). More real-life example would be to check authorization header presence in `jsonrpc-core` middleware.

It's also used by our PubSub mechanism (also a custom extension), where metadata contains a pipe allowing the method handlers to send notifications to the clients.

`jsonrpc_core::IoHandler` is a pure implementation of JSONRPC spec. This extension is available in `jsonrpc_core::MetaIoHandler`.

Let me know if you need any more detailed info on this.

**Comment by greenpdx** [comment]
I need some help with Metadata, If I could get just a simple template I could do it. This is what I have now

''' rust
extern crate jsonrpc_core;
extern crate jsonrpc_http_server;

use jsonrpc_core::*;
use jsonrpc_http_server::{ServerBuilder, DomainsValidation, AccessControlAllowOrigin, RestApi, MetaExtractor};

#[derive(Clone, Default)]
struct Meta(usize);
impl MetaExtractor<Meta> for Meta {
    fn read_metadata(&self, req: &server::Request) -> Meta {
        (*self)(req)
    }
}

fn main() {
    let mut io = MetaIoHandler::default();
    io.add_method_with_meta("say_hello", |_parm: Params, meta: Meta| {
        println!("Hello {:?}", _parm );
        futures::finished(Value::String(format!("hello: {}", meta.0)))
        //Ok(Value::String("hello".to_string()))
    });

    let _server = ServerBuilder::new(io)
        .threads(3)
        .rest_api(RestApi::Unsecure)
        .meta_extractor(Meta::read_metadata())
        .cors(DomainsValidation::AllowOnly(vec![AccessControlAllowOrigin::Any]))
        .start_http(&"0.0.0.0:3030".parse().unwrap())
        .expect("unable to start");

    _server.wait();
}

'''

**Comment by tomusdrw** [comment]
@greenpdx Please have a look at the example I added here: https://github.com/paritytech/jsonrpc/pull/219

**Comment by greenpdx** [comment]
I am testing it using Vue, send it a json request with params
{ jsonrpc: 2.0, method: say_hello, params: [23, 16], id: 1}

but I get this error
{ "jsonrpc": "2.0", "error": { "code": -32600, "message": "Invalid request" }, "id": 1 }

**Comment by tomusdrw** [comment]
Can you make sure that the request is a properly formed JSON?
Can you also run your rust server with `RUST_LOG="rpc=trace"` (and please initialize `env_logger`)? It wil then display the trace data.

**Comment by greenpdx** [comment]
Yes, it was me. I was added a header for cors that I did not need.  I am changing the  Header from Authorization to Host, because I wanted to use an existing request header. it works
thanks

**Comment by tomusdrw** [comment]
Perfect! Glad it works :)

**Comment by greenpdx** [comment]
Now I am trying it with remote_addr.
next I want to add a meta and middleware to it
